examples/hubert/dataset/hubert_dataset.py

```diff
@@ -78,56 +89,59 @@ def _get_buckets(
         """
         buckets = {}
 
-        boundaries = [min_len - 1]
         interval = (max_len - min_len) // num_buckets
-        for i in range(1, num_buckets):
-            boundaries.append(min_len + i * interval)
-        boundaries.append(max_len + 1)
-        bucket_ids = torch.bucketize(torch.tensor(lengths), torch.tensor(boundaries))
-        for i in indices:
+        boundaries = torch.linspace(min_len, max_len, interval)
+        bucket_ids = torch.bucketize(torch.tensor(lengths), boundaries)
+        for i in range(bucket_ids.size(0)):
             bucket_id = bucket_ids[i]
             if bucket_id in buckets:
                 buckets[bucket_id].append(i)
             else:
                 buckets[bucket_id] = [i]
         for k in buckets:
-            if self.shuffle:
-                random.shuffle(buckets[k])
             buckets[k] = torch.as_tensor(buckets[k], dtype=torch.int)
         return buckets
 
     def __iter__(self) -> Iterator[List[int]]:
+        if self.shuffle:
+            for k in self.buckets:
+                self.buckets[k] = self.buckets[k][torch.randperm(self.buckets[k].size(0))]
+
         iter_list = []
         total_len = 0
         batch = []
-        if self.max_token_count:
-            for k in self.buckets.keys():
-                for i in range(self.buckets[k].size(0)):
-                    index = self.buckets[k][i]
-                    if total_len > self.max_token_count:
-                        iter_list.append(batch)
-                        batch = [index]
-                        total_len = self.lengths[index]
-                    else:
-                        batch.append(index)
+        for k in sorted(self.buckets):
+            for i in range(self.buckets[k].size(0)):
+                index = int(self.buckets[k][i])
+                if self.max_token_count is not None:
+                    if total_len + self.lengths[index] <= self.max_token_count:
+                        batch.append(self.indices[index])
                         total_len += self.lengths[index]
-        else:
-            for k in self.buckets.keys():
-                for i in range(self.buckets[k].size(0)):
-                    index = self.buckets[k][i]
+                    else:
+                        iter_list.append(batch)
+                        batch = []
+                        total_len = 0
+                else:
                     if total_len == self.batch_size:
                         iter_list.append(batch)
-                        batch = [index]
+                        batch = [self.indices[index]]
                         total_len = 1
                     else:
-                        batch.append(index)
+                        batch.append(self.indices[index])
                         total_len += 1
+            if len(batch) > 0 and (self.drop_last or self.max_token_count):
```

Comment:
`self.drop_last` being `True` should result in the last batch being dropped if it's partially filled and we're batching by sample count — i don't think this is captured properly here